### ISSUE TITLE
Assertion Failure When Closing Database

### ISSUE THREAD

**Issue by cmumford** [google/leveldb#142]
Original [issue 136](https://code.google.com/p/leveldb/issues/detail?id=136) created by stephen.p.morgan on 2013-01-03T19:33:17.000Z:

<b>What steps will reproduce the problem?</b>
1. open array of dbs, &quot;db[n]&quot;
2. close db using &quot;delete db[n]&quot;

<b>What is the expected output? What do you see instead?</b>
Expect db to close. Instead, see assertion failure:

```
main: db/version_set.cc:715: leveldb::VersionSet::~VersionSet(): Assertion 'dummy_versions_.next_ == &dummy_versions_' failed.
```

<b>What version of the product are you using? On what operating system?</b>
1.6.0 on Ubuntu Linux kernel 3.2.1, x86_64

<b>Please provide any additional information below.</b>

**Comment by aroth-fastprotect** [comment]
With the follow piece of code i could trigger this assertion (when the DB instance is deleted):

    #include <string>
    #include <iostream>
    #include <leveldb/options.h>
    #include <leveldb/db.h>
    #include <leveldb/env.h>


    void dump(leveldb::DB* db, std::ostream & os)
    {
        leveldb::Iterator* i = db->NewIterator(leveldb::ReadOptions());
        for(i->SeekToFirst(); i->Valid(); i->Next())
        {
            std::string key = i->key().ToString();
            //if (key.compare(0, 2, "b!", 2) == 0)
            {
                std::string value = i->value().ToString();
                os << "      " << key.data() << ": " << value.data() << std::endl;
            }
        }
    }

    class ostream_logger : public leveldb::Logger
    {
    public:
        ostream_logger(std::ostream & os) : _os(os) { }
        virtual ~ostream_logger() {}

        // Write an entry to the log file with the specified format.
        virtual void Logv(const char* format, va_list ap)
        {
            char buf[1024];
            vsprintf(buf, format, ap);
            _os << buf << std::endl;
        }

    private:
        // No copying allowed
        ostream_logger(const ostream_logger&);
        void operator=(const ostream_logger&);
    private:
        std::ostream & _os;
    };


    int main(int argc, const char * argv[])
    {
        std::string rootPath = argv[1];

        ostream_logger logger(std::cout);

        int ret = 0;
        leveldb::DB* _db;
        leveldb::Options dboptions;
        dboptions.info_log = &logger;
        dboptions.create_if_missing = false;
        dboptions.error_if_exists   = false;

        leveldb::Status status;

        //status = leveldb::RepairDB(rootPath, dboptions);

        status = leveldb::DB::Open(dboptions, rootPath, &_db);
        if ( status.ok() )
        {
            dump(_db, std::cout);
        }
        delete _db;


        return ret;
    }

I'm using version 1.19 of leveldb on Win32.

**Comment by cmumford** [comment]
I had to switch create_if_missing to true (at least once) in order to get a test directory. After which I get this stack trace:

    F1019 10:18:04.176285   24059 logging.cc:2148] assert.h assertion failed at third_party/leveldb/db/version_set.cc:789 in leveldb::VersionSet::~VersionSet(): dummy_versions_.next_ == &dummy_versions_
    *** Check failure stack trace: ***
        @     0x7f3aba28b2ea  base_logging::LogMessage::SendToLog()
        @     0x7f3aba28bc9a  base_logging::LogMessage::Flush()
        @     0x7f3aba28e315  LogMessageFatal::~LogMessageFatal()
        @     0x7f3aba2892c7  __assert_fail
        @     0x7f3aba5c09fc  leveldb::VersionSet::~VersionSet()
        @     0x7f3aba5a1f9b  leveldb::DBImpl::~DBImpl()
        @     0x7f3aba5a21e9  leveldb::DBImpl::~DBImpl()
        @           0x407bd1  main
        @     0x7f3ab92bace8  __libc_start_main
        @           0x407499  _start

**Comment by aroth-fastprotect** [comment]
Yeah, i forgot. I used an existing database. But the result you've got is exactly the same for me.

Have you had a chance to look into this issue?

**Comment by aroth-arsoft** [comment]
I've just tried the latest version of LevelDB, but i ran into this assert again. 

From what i understand some Version object is still stuck in the VersionSet because it's still reference at some point.

**Comment by marcinja** [comment]
I came across the same assert failure and found that other instances of `leveldb::Iterator` in the codebase I'm working in are wrapped in `std::unique_ptr`.

I tried doing the same and the failure disappeared. Alternatively, calling `delete iter;` also solved the problem. 

I don't know anything about leveldb internals, but I guess it doesn't like having open iterators laying around.

**Comment by aroth-arsoft** [comment]
Yeah, i can confirm that deleting the iterator fixed the issue; no assert anymore. So this issue can be closed!?

**Comment by houzhj58** [comment]
just delete **dummy_versions_.next_**？

**Comment by houzhj58** [comment]
Maybe you new this    **leveldb::Iterator* it = db->NewIterator()**
Then just delete the **it**
It's OK.
